File: src/blogerl_app.erl
%%%-------------------------------------------------------------------
%% @doc blogerl public API
%% @end
%%%-------------------------------------------------------------------

-module(blogerl_app).

-behaviour(application).

-export([start/2, stop/1]).

start(_StartType, _StartArgs) ->
    blogerl_sup:start_link(blogerl_storage_dets),
    setup_cowboy(blogerl_storage_dets).

setup_cowboy(Storage_mod) ->
    Dispatch = cowboy_router:compile([
        {'_', [
            {"/:title", blogerl_index_h, #{storage_mod => Storage_mod, storage_pid => Storage_mod}},
            {"/", blogerl_index_h, #{storage_mod => Storage_mod, storage_pid => Storage_mod}}
        ]}
    ]),
    {ok, _} = cowboy:start_clear(http, [{port, 8080}], #{
        env => #{dispatch => Dispatch}
    }).
stop(_State) ->
    ok.

%% internal functions

File: lib/blogerl_app.ex
defmodule Blogerl.Application do
  use Application

  def start(_type, _args) do
    Blogerl.Supervisor.start_link(name: Blogerl.Supervisor)
    dispatcher = :cowboy_router.compile([
      # {HostMatch, [{PathMatch, Handler, InitialState}]
      { :_,
        [
          {"/", Blogerl.HttpHandler, []},
          {"/:title", Blogerl.HttpHandler, []},
        ]
      }
    ])
    :cowboy.start_clear(:my_http_listener,
    [{:port, 8080}],
    %{:env => %{:dispatch => dispatcher}})
  end
end
